File: quickstart/commands/certs.py
"""qs certs — Thin wrapper around gen_certs.sh with certificate verification."""

from __future__ import annotations

import argparse
import subprocess

from quickstart import runner
from quickstart.config import ProjectConfig
from quickstart.preflight import require


def _gen_script(config: ProjectConfig) -> str:
    return str(config.cloud_dir / config.gen_certs_script)


def _certs_dir(config: ProjectConfig):
    return config.repo_root / config.certs_dir_name


def _run_gen(config: ProjectConfig, gen_args: list[str], *, verbose: bool, dry_run: bool) -> int:
    """Run gen_certs.sh with the given arguments. Returns 0 on success."""
    try:
        runner.run(
            [_gen_script(config)] + gen_args,
            cwd=config.cloud_dir, verbose=verbose, dry_run=dry_run,
        )
        return 0
    except subprocess.CalledProcessError as e:
        runner.error(f"gen_certs.sh {' '.join(gen_args)} failed: {e}")
        return 1


def cmd_ca(args: argparse.Namespace, config: ProjectConfig) -> int:
    return _run_gen(config, ["ca"], verbose=args.verbose, dry_run=args.dry_run)


def cmd_device(args: argparse.Namespace, config: ProjectConfig) -> int:
    gen_args = ["device", args.name]
    if args.days:
        gen_args.append(str(args.days))
    return _run_gen(config, gen_args, verbose=args.verbose, dry_run=args.dry_run)


def cmd_server(args: argparse.Namespace, config: ProjectConfig) -> int:
    gen_args = ["server", args.name, args.host_ip]
    if args.days:
        gen_args.append(str(args.days))
    return _run_gen(config, gen_args, verbose=args.verbose, dry_run=args.dry_run)


def cmd_list(args: argparse.Namespace, config: ProjectConfig) -> int:
    return _run_gen(config, ["list"], verbose=args.verbose, dry_run=args.dry_run)


def cmd_verify(args: argparse.Namespace, config: ProjectConfig) -> int:
    """Verify the mTLS certificate chain using OpenSSL."""
    verbose = args.verbose
    dry_run = args.dry_run
    certs = _certs_dir(config)

    ca_pem = str(certs / "ca" / "ca.pem")
    server_pem = str(certs / "servers" / args.server_name / "server.pem")
    client_pem = str(certs / "devices" / args.device_name / "client.pem")

    all_ok = True

    def _check(label: str, cmd: list[str], *, expect_ok: bool = True) -> bool:
        """Run an openssl command. Returns True if check passed."""
        runner.step(label)
        try:
            result = runner.run(cmd, capture=True, verbose=verbose, dry_run=dry_run)
            if dry_run:
                return True
            if expect_ok and "OK" not in result.stdout:
                runner.error(f"  FAIL: {result.stdout.strip()}")
                return False
            runner.success(f"  {result.stdout.strip()}")
            return True
        except subprocess.CalledProcessError as e:
            runner.error(f"  FAIL: {e}")
            return False

    # 1. Verify CA is valid (self-signed)
    if not _check("Verify root CA", ["openssl", "verify", "-CAfile", ca_pem, ca_pem]):
        all_ok = False

    # 2. Verify server cert against CA
    if not _check("Verify server cert against CA", ["openssl", "verify", "-CAfile", ca_pem, server_pem]):
        all_ok = False

    # 3. Verify client cert against CA
    if not _check("Verify client cert against CA", ["openssl", "verify", "-CAfile", ca_pem, client_pem]):
        all_ok = False

    # 4. Inspect server cert SAN
    _check(
        "Server cert SAN",
        ["openssl", "x509", "-in", server_pem, "-noout", "-ext", "subjectAltName"],
        expect_ok=False,
    )

    # 5. Inspect client cert subject DN
    _check(
        "Client cert subject",
        ["openssl", "x509", "-in", client_pem, "-noout", "-subject"],
        expect_ok=False,
    )

    # 6. Verify client cert has clientAuth Extended Key Usage
    runner.step("Client cert Extended Key Usage")
    try:
        result = runner.run(
            ["openssl", "x509", "-in", client_pem, "-noout", "-ext", "extendedKeyUsage"],
            capture=True, verbose=verbose, dry_run=dry_run,
        )
        if not dry_run:
            eku_output = result.stdout.strip()
            if "clientAuth" in eku_output:
                runner.success(f"  {eku_output}")
            else:
                runner.error(f"  FAIL: clientAuth not found in Extended Key Usage: {eku_output}")
                all_ok = False
    except subprocess.CalledProcessError as e:
        runner.error(f"  FAIL: could not read Extended Key Usage: {e}")
        all_ok = False

    # 7. Check expiry on all three
    runner.step("Certificate expiry dates")
    for label, cert in [("CA", ca_pem), ("Server", server_pem), ("Client", client_pem)]:
        try:
            result = runner.run(
                ["openssl", "x509", "-in", cert, "-noout", "-enddate"],
                capture=True, verbose=verbose, dry_run=dry_run,
            )
            if not dry_run:
                expiry = result.stdout.strip().removeprefix("notAfter=")
                runner.success(f"  {label}: expires {expiry}")
        except subprocess.CalledProcessError as e:
            runner.error(f"  {label} expiry check failed: {e}")
            all_ok = False

    if all_ok:
        runner.success("\nAll checks passed.")
        return 0
    else:
        runner.error("\nSome checks failed.")
        return 1


def cmd_handshake(args: argparse.Namespace, config: ProjectConfig) -> int:
    """Simulate an mTLS handshake against a running server."""
    verbose = args.verbose
    dry_run = args.dry_run
    certs = _certs_dir(config)

    ca_pem = str(certs / "ca" / "ca.pem")
    client_pem = str(certs / "devices" / args.device_name / "client.pem")
    client_key = str(certs / "devices" / args.device_name / "client.key")
    target = f"{args.host}:{args.port}"

    all_ok = True

    # 1. Full mTLS handshake (client cert presented)
    runner.step(f"mTLS handshake to {target}")
    mtls_cmd = [
        "openssl", "s_client", "-connect", target,
        "-cert", client_pem,
        "-key", client_key,
        "-CAfile", ca_pem,
    ]
    if args.tls_version:
        mtls_cmd.append(f"-{args.tls_version}")
    try:
        result = runner.run(
            mtls_cmd,
            capture=True, verbose=verbose, dry_run=dry_run,
        )
        if not dry_run:
            if "Verify return code: 0 (ok)" in result.stdout:
                runner.success("  Handshake OK — Verify return code: 0 (ok)")
            else:
                # Extract the verify return code line for diagnostics
                for line in result.stdout.splitlines():
                    if "Verify return code:" in line:
                        runner.error(f"  FAIL: {line.strip()}")
                        break
                else:
                    runner.error("  FAIL: could not find verify return code in output")
                all_ok = False
    except subprocess.CalledProcessError as e:
        runner.error(f"  FAIL: mTLS handshake failed: {e}")
        all_ok = False

    # 2. Without client cert — should be rejected if mTLS is enforced
    if args.test_no_client:
        runner.step(f"Connecting without client cert to {target} (should fail if mTLS enforced)")
        no_client_cmd = [
            "openssl", "s_client", "-connect", target,
            "-CAfile", ca_pem,
        ]
        if args.tls_version:
            no_client_cmd.append(f"-{args.tls_version}")
        try:
            result = runner.run(
                no_client_cmd,
                capture=True, check=False, verbose=verbose, dry_run=dry_run,
            )
            if not dry_run:
                if "Verify return code: 0 (ok)" in result.stdout:
                    runner.error("  WARN: server accepted connection without client cert — mTLS may not be enforced")
                else:
                    runner.success("  Server rejected connection without client cert (mTLS enforced)")
        except subprocess.CalledProcessError:
            runner.success("  Server rejected connection without client cert (mTLS enforced)")

    if all_ok:
        runner.success("\nHandshake checks passed.")
        return 0
    else:
        runner.error("\nHandshake checks failed.")
        return 1


def run_certs(args: argparse.Namespace, config: ProjectConfig) -> int:
    """Dispatch to the appropriate certs subcommand."""
    require(["openssl"])
    dispatch = {
        "ca": cmd_ca,
        "device": cmd_device,
        "server": cmd_server,
        "list": cmd_list,
        "verify": cmd_verify,
        "handshake": cmd_handshake,
    }
    return dispatch[args.certs_command](args, config)


def register_subparser(sub: argparse._SubParsersAction) -> None:
    """Register the certs command and its subcommands on the main CLI parser."""
    certs_p = sub.add_parser("certs", help="Generate and verify mTLS certificates")
    certs_sub = certs_p.add_subparsers(dest="certs_command", required=True)

    certs_sub.add_parser("ca", help="Initialize the root CA")

    device_p = certs_sub.add_parser("device", help="Issue a device client certificate")
    device_p.add_argument("name", help="Device name (e.g. esp32)")
    device_p.add_argument("--days", type=int, default=None, help="Validity in days")

    server_p = certs_sub.add_parser("server", help="Issue a server/TLS certificate")
    server_p.add_argument("name", help="Server name (e.g. caddy)")
    server_p.add_argument("--host-ip", default="127.0.0.1", help="SAN IP (default: 127.0.0.1)")
    server_p.add_argument("--days", type=int, default=None, help="Validity in days")

    certs_sub.add_parser("list", help="List all issued certificates")

    verify_p = certs_sub.add_parser("verify", help="Verify the mTLS certificate chain")
    verify_p.add_argument("--device-name", default="esp32", help="Device cert to verify (default: esp32)")
    verify_p.add_argument("--server-name", default="caddy", help="Server cert to verify (default: caddy)")

    hs_p = certs_sub.add_parser("handshake", help="Simulate mTLS handshake against a running server")
    hs_p.add_argument("--host", default="localhost", help="Server host (default: localhost)")
    hs_p.add_argument("--port", default="443", help="Server port (default: 443)")
    hs_p.add_argument("--device-name", default="esp32", help="Device cert to use (default: esp32)")
    hs_p.add_argument("--server-name", default="caddy", help="Server cert name (default: caddy)")
    hs_p.add_argument("--tls-version", default="tls1_3", help="TLS version flag (default: tls1_3)")
    hs_p.add_argument("--test-no-client", action="store_true", help="Also test without client cert to confirm mTLS is enforced")

File: supervictor/cli/src/commands/certs.rs
use crate::config::ProjectConfig;
use crate::error::CliError;
use crate::preflight;
use crate::runner::{self, RunOptions, Runner};

/// Arguments for the `qs certs` command.
pub struct CertsArgs {
    /// Enable verbose output.
    pub verbose: bool,
    /// Print commands without executing.
    pub dry_run: bool,
    /// The certificate subcommand to execute.
    pub command: CertsCommand,
}

/// Certificate management subcommands.
pub enum CertsCommand {
    /// Generate a new root CA.
    Ca,
    /// Generate a device client certificate.
    Device {
        /// Device identity name (used as directory and CN).
        name: String,
        /// Optional validity period in days.
        days: Option<u32>,
    },
    /// Generate a server certificate with SAN.
    Server {
        /// Server name (used as directory).
        name: String,
        /// IP or hostname for the SAN extension.
        host_ip: String,
        /// Optional validity period in days.
        days: Option<u32>,
    },
    /// List all generated certificates.
    List,
    /// Verify a device and server cert chain against the CA.
    Verify {
        /// Device whose client cert to verify.
        device_name: String,
        /// Server whose cert to verify.
        server_name: String,
    },
    /// Test a live TLS handshake against a remote server.
    Handshake {
        /// Target hostname.
        host: String,
        /// Target port.
        port: String,
        /// Device name whose client cert to present.
        device_name: String,
        /// Optional TLS version flag (e.g. "tls1_2").
        tls_version: Option<String>,
        /// Also test connecting without a client cert.
        test_no_client: bool,
    },
}

/// Dispatch the `qs certs` subcommand.
pub fn run_certs(
    args: &CertsArgs,
    config: &ProjectConfig,
    r: &dyn Runner,
) -> Result<i32, CliError> {
    preflight::require(&["openssl"], false, r)?;

    match &args.command {
        CertsCommand::Ca => run_gen(config, &["ca"], args.verbose, args.dry_run, r),
        CertsCommand::Device { name, days } => {
            let mut gen_args = vec!["device", name.as_str()];
            let days_str;
            if let Some(d) = days {
                days_str = d.to_string();
                gen_args.push(&days_str);
            }
            run_gen(config, &gen_args, args.verbose, args.dry_run, r)
        }
        CertsCommand::Server {
            name,
            host_ip,
            days,
        } => {
            let mut gen_args = vec!["server", name.as_str(), host_ip.as_str()];
            let days_str;
            if let Some(d) = days {
                days_str = d.to_string();
                gen_args.push(&days_str);
            }
            run_gen(config, &gen_args, args.verbose, args.dry_run, r)
        }
        CertsCommand::List => run_gen(config, &["list"], args.verbose, args.dry_run, r),
        CertsCommand::Verify {
            device_name,
            server_name,
        } => cmd_verify(
            config,
            device_name,
            server_name,
            args.verbose,
            args.dry_run,
            r,
        ),
        CertsCommand::Handshake {
            host,
            port,
            device_name,
            tls_version,
            test_no_client,
        } => cmd_handshake(
            config,
            &HandshakeArgs {
                host,
                port,
                device_name,
                tls_version: tls_version.as_deref(),
                test_no_client: *test_no_client,
                verbose: args.verbose,
                dry_run: args.dry_run,
            },
            r,
        ),
    }
}

fn run_gen(
    config: &ProjectConfig,
    gen_args: &[&str],
    verbose: bool,
    dry_run: bool,
    r: &dyn Runner,
) -> Result<i32, CliError> {
    let script = config.gen_certs_script_path();
    let script_str = script.to_string_lossy().to_string();
    let mut cmd: Vec<&str> = vec![&script_str];
    cmd.extend_from_slice(gen_args);

    match r.run(
        &cmd,
        &RunOptions {
            cwd: Some(config.repo_root.clone()),
            verbose,
            dry_run,
            ..Default::default()
        },
    ) {
        Ok(_) => Ok(0),
        Err(e) => {
            runner::error(&format!(
                "gen_certs.sh {} failed: {}",
                gen_args.join(" "),
                e
            ));
            Ok(1)
        }
    }
}

fn cmd_verify(
    config: &ProjectConfig,
    device_name: &str,
    server_name: &str,
    verbose: bool,
    dry_run: bool,
    r: &dyn Runner,
) -> Result<i32, CliError> {
    let certs = config.certs_dir();
    let ca_pem = certs.join("ca/ca.pem");
    let server_pem = certs.join(format!("servers/{}/server.pem", server_name));
    let client_pem = certs.join(format!("devices/{}/client.pem", device_name));
    let ca_str = ca_pem.to_string_lossy().to_string();
    let server_str = server_pem.to_string_lossy().to_string();
    let client_str = client_pem.to_string_lossy().to_string();

    let opts = RunOptions {
        capture: true,
        verbose,
        dry_run,
        ..Default::default()
    };

    let mut all_ok = true;

    // 1. Verify CA
    runner::step("Verify root CA");
    match r.run(&["openssl", "verify", "-CAfile", &ca_str, &ca_str], &opts) {
        Ok(result) if dry_run || result.stdout.contains("OK") => {
            runner::success(&format!("  {}", result.stdout.trim()));
        }
        Ok(result) => {
            runner::error(&format!("  FAIL: {}", result.stdout.trim()));
            all_ok = false;
        }
        Err(e) => {
            runner::error(&format!("  FAIL: {}", e));
            all_ok = false;
        }
    }

    // 2. Verify server cert
    runner::step("Verify server cert against CA");
    match r.run(
        &["openssl", "verify", "-CAfile", &ca_str, &server_str],
        &opts,
    ) {
        Ok(result) if dry_run || result.stdout.contains("OK") => {
            runner::success(&format!("  {}", result.stdout.trim()));
        }
        Ok(result) => {
            runner::error(&format!("  FAIL: {}", result.stdout.trim()));
            all_ok = false;
        }
        Err(e) => {
            runner::error(&format!("  FAIL: {}", e));
            all_ok = false;
        }
    }

    // 3. Verify client cert
    runner::step("Verify client cert against CA");
    match r.run(
        &["openssl", "verify", "-CAfile", &ca_str, &client_str],
        &opts,
    ) {
        Ok(result) if dry_run || result.stdout.contains("OK") => {
            runner::success(&format!("  {}", result.stdout.trim()));
        }
        Ok(result) => {
            runner::error(&format!("  FAIL: {}", result.stdout.trim()));
            all_ok = false;
        }
        Err(e) => {
            runner::error(&format!("  FAIL: {}", e));
            all_ok = false;
        }
    }

    // 4. Server cert SAN
    runner::step("Server cert SAN");
    if let Ok(result) = r.run(
        &[
            "openssl",
            "x509",
            "-in",
            &server_str,
            "-noout",
            "-ext",
            "subjectAltName",
        ],
        &RunOptions {
            check: false,
            ..opts.clone()
        },
    ) {
        runner::success(&format!("  {}", result.stdout.trim()));
    }

    // 5. Client cert subject
    runner::step("Client cert subject");
    if let Ok(result) = r.run(
        &["openssl", "x509", "-in", &client_str, "-noout", "-subject"],
        &RunOptions {
            check: false,
            ..opts.clone()
        },
    ) {
        runner::success(&format!("  {}", result.stdout.trim()));
    }

    // 6. Client cert Extended Key Usage
    runner::step("Client cert Extended Key Usage");
    match r.run(
        &[
            "openssl",
            "x509",
            "-in",
            &client_str,
            "-noout",
            "-ext",
            "extendedKeyUsage",
        ],
        &RunOptions {
            check: false,
            ..opts.clone()
        },
    ) {
        Ok(result) if !dry_run => {
            if result.stdout.contains("clientAuth") {
                runner::success(&format!("  {}", result.stdout.trim()));
            } else {
                runner::error(&format!(
                    "  FAIL: clientAuth not found in Extended Key Usage: {}",
                    result.stdout.trim()
                ));
                all_ok = false;
            }
        }
        Ok(result) => runner::success(&format!("  {}", result.stdout.trim())),
        Err(e) => {
            runner::error(&format!("  FAIL: could not read Extended Key Usage: {}", e));
            all_ok = false;
        }
    }

    // 7. Expiry dates
    runner::step("Certificate expiry dates");
    for (label, cert_path) in [
        ("CA", &ca_str),
        ("Server", &server_str),
        ("Client", &client_str),
    ] {
        match r.run(
            &["openssl", "x509", "-in", cert_path, "-noout", "-enddate"],
            &RunOptions {
                check: false,
                ..opts.clone()
            },
        ) {
            Ok(result) if !dry_run => {
                let expiry = result
                    .stdout
                    .trim()
                    .strip_prefix("notAfter=")
                    .unwrap_or(result.stdout.trim());
                runner::success(&format!("  {}: expires {}", label, expiry));
            }
            Ok(_) => {}
            Err(e) => {
                runner::error(&format!("  {} expiry check failed: {}", label, e));
                all_ok = false;
            }
        }
    }

    if all_ok {
        runner::success("\nAll checks passed.");
        Ok(0)
    } else {
        runner::error("\nSome checks failed.");
        Ok(1)
    }
}

struct HandshakeArgs<'a> {
    host: &'a str,
    port: &'a str,
    device_name: &'a str,
    tls_version: Option<&'a str>,
    test_no_client: bool,
    verbose: bool,
    dry_run: bool,
}

fn cmd_handshake(
    config: &ProjectConfig,
    ha: &HandshakeArgs<'_>,
    r: &dyn Runner,
) -> Result<i32, CliError> {
    let host = ha.host;
    let port = ha.port;
    let device_name = ha.device_name;
    let tls_version = ha.tls_version;
    let test_no_client = ha.test_no_client;
    let verbose = ha.verbose;
    let dry_run = ha.dry_run;
    let certs = config.certs_dir();
    let ca_pem = certs.join("ca/ca.pem");
    let client_pem = certs.join(format!("devices/{}/client.pem", device_name));
    let client_key = certs.join(format!("devices/{}/client.key", device_name));
    let ca_str = ca_pem.to_string_lossy().to_string();
    let client_pem_str = client_pem.to_string_lossy().to_string();
    let client_key_str = client_key.to_string_lossy().to_string();
    let target = format!("{}:{}", host, port);

    let opts = RunOptions {
        capture: true,
        verbose,
        dry_run,
        check: false,
        ..Default::default()
    };

    let mut all_ok = true;

    // 1. Full mTLS handshake
    runner::step(&format!("mTLS handshake to {}", target));
    let mut mtls_cmd = vec![
        "openssl",
        "s_client",
        "-connect",
        &target,
        "-cert",
        &client_pem_str,
        "-key",
        &client_key_str,
        "-CAfile",
        &ca_str,
    ];
    let tls_flag = tls_version.map(|ver| format!("-{}", ver));
    if let Some(ref flag) = tls_flag {
        mtls_cmd.push(flag);
    }

    match r.run(&mtls_cmd, &opts) {
        Ok(result) if !dry_run => {
            if result.stdout.contains("Verify return code: 0 (ok)") {
                runner::success("  Handshake OK — Verify return code: 0 (ok)");
            } else {
                let diag = result
                    .stdout
                    .lines()
                    .find(|l| l.contains("Verify return code:"))
                    .unwrap_or("could not find verify return code in output");
                runner::error(&format!("  FAIL: {}", diag.trim()));
                all_ok = false;
            }
        }
        Ok(_) => {}
        Err(e) => {
            runner::error(&format!("  FAIL: mTLS handshake failed: {}", e));
            all_ok = false;
        }
    }

    // 2. Without client cert
    if test_no_client {
        runner::step(&format!(
            "Connecting without client cert to {} (should fail if mTLS enforced)",
            target
        ));
        let mut no_client_cmd = vec![
            "openssl", "s_client", "-connect", &target, "-CAfile", &ca_str,
        ];
        if let Some(ref flag) = tls_flag {
            no_client_cmd.push(flag);
        }

        match r.run(&no_client_cmd, &opts) {
            Ok(result) if !dry_run => {
                if result.stdout.contains("Verify return code: 0 (ok)") {
                    runner::error(
                        "  WARN: server accepted connection without client cert — mTLS may not be enforced",
                    );
                } else {
                    runner::success(
                        "  Server rejected connection without client cert (mTLS enforced)",
                    );
                }
            }
            Ok(_) => {}
            Err(_) => {
                runner::success("  Server rejected connection without client cert (mTLS enforced)");
            }
        }
    }

    if all_ok {
        runner::success("\nHandshake checks passed.");
        Ok(0)
    } else {
        runner::error("\nHandshake checks failed.");
        Ok(1)
    }
}
